src/app/battle.service.ts

```diff
@@ -178,9 +188,18 @@ export class BattleService {
       [input.playerFleet, playerShips, ret.playerLost],
       [input.enemyFleet, enemyShip, ret.enemyLost]
     ];
+
     retArr.forEach(arr => {
+      const fleetCount = {};
+
+      // Count and free the ships object
+      for (const ship of arr[1]){
+        fleetCount[ship.id] = fleetCount[ship.id] ? fleetCount[ship.id] + 1 : 1;
+        ship.free();
+      }
```

Comment:
A single loop to count the amount of ship per type is better than using filter multiple times.
Plus we need to free the ship, no better place to do it than here.